### ISSUE TITLE
DPO/ORPO preference-based refinement after SFT

### ISSUE THREAD

**Issue by NP-compete** [NP-compete/smithery#16]
## Problem

Supervised fine-tuning (SFT) teaches the model what good tool calls look like. But it doesn't teach the model to *prefer* good tool calls over subtly wrong ones. Direct Preference Optimization (DPO) or Odds Ratio Preference Optimization (ORPO) can squeeze out the last 3-5% accuracy by training on paired examples: "this tool call is better than that one."

### Preference Data Generation

After SFT, generate preference pairs:

1. Run the SFT model on the test set
2. For each example where the model gets it wrong (or partially right):
   - **Chosen**: the correct tool call (from ground truth)
   - **Rejected**: the model's incorrect prediction

3. Also generate synthetic "hard negatives":
   - Right tool, wrong argument value
   - Right tool, missing required argument
   - Wrong tool but plausible (similar name/description)
   - Right tool call but with extra hallucinated arguments

### Training

\`\`\`python
from smithery.train import PreferenceTrainer

trainer = PreferenceTrainer(
    method="dpo",          # or "orpo"
    base_model="./output/my-sft-agent",  # start from SFT checkpoint
    beta=0.1,              # DPO temperature
)
result = trainer.train(preference_dataset)
\`\`\`

### Method Comparison

| Method | Pros | Cons |
|---|---|---|
| **DPO** | Well-studied, stable, TRL has built-in support | Requires reference model (2x memory) |
| **ORPO** | No reference model needed (1x memory), simpler | Newer, less battle-tested |

Recommend: support both, default to ORPO for memory efficiency on consumer GPUs.

### Expected Improvement

Based on literature:
- Tool selection: +2-4% accuracy on edge cases
- Parameter extraction: +0.02-0.05 F1 on ambiguous inputs
- Safety: significantly better refusal calibration (fewer false refusals)

## Acceptance Criteria

- [ ] \`PreferenceTrainer\` class supporting DPO and ORPO
- [ ] Preference data generation from SFT model failures
- [ ] Hard negative generation (wrong tool, wrong args, extra args)
- [ ] Recipe YAML configs for DPO/ORPO
- [ ] CLI: \`smithery train --method dpo --data ./preferences.jsonl\`
- [ ] Benchmark: SFT vs SFT+DPO accuracy comparison
- [ ] Works on a single consumer GPU (ORPO default for memory)
- [ ] Documentation on when to use DPO vs ORPO

## Dependencies

- Depends on #8 (SFT training — DPO/ORPO runs after)
- Depends on #9 (evaluation — to generate failure data)
- TRL has built-in DPOTrainer

## References

- [DPO paper](https://arxiv.org/abs/2305.18290) — Direct Preference Optimization
- [ORPO paper](https://arxiv.org/abs/2403.07691) — Odds Ratio Preference Optimization
- [TRL DPOTrainer](https://huggingface.co/docs/trl/main/en/dpo_trainer)

## Estimated Effort

Medium (5-7 days). The TRL integration is straightforward. Generating high-quality preference pairs requires experimentation.
